lms/static/completion/js/ViewedEvent.js

```diff
@@ -0,0 +1,158 @@
+function throttle(fn, wait) {
+  let time = Date.now();
+  function delay() {
+    const now = Date.now();
+    if (time + wait < now) {
+      time = now;
+      fn();
+    }
+  }
+  return delay;
+}
+
+export class ElementViewing {
+  constructor(el, completeAfterMs) {
+    this.el = el;
+    this.completeAfterMs = completeAfterMs;
+    this.topSeen = false;
+    this.bottomSeen = false;
+    this.seenForMs = 0;
+    this.lastSeen = undefined;
+    this.complete = false;
+  }
+
+  getBoundingRect() {
+    return this.el.getBoundingClientRect();
+  }
+
+  pauseTracking() {
+    this.lastSeen = undefined;
+  }
+
+  markTopSeen() {
+    this.topSeen = true;
+  }
+
+  markBottomSeen() {
+    this.bottomSeen = true;
+  }
+
+  hasBeenViewed() {
+    return this.topSeen && this.bottomSeen && (this.seenForMs >= this.completeAfterMs);
+  }
+
+  handleVisible(now) {
+    if (this.lastSeen) {
+      this.seenForMs += (now - this.lastSeen);
+    }
+    this.lastSeen = now;
+  }
+
+  sendEvent(callback) {
+    callback(this.el, {
+      elementHasBeenViewed: this.hasBeenViewed(),
+    });
+  }
+}
+
+
+export class ViewedEventTracker {
+  /**
+   * When the top or bottom of an element is
+   * first viewed, or the entire element is viewed for a specified amount
+   * of time, the callback is called, passing the element that was viewed,
+   * and an event object having the following field:
+   *
+   * *   hasBeenViewed (bool): true if all the conditions for being
+   *     considered "viewed" have been met.
+   */
+  constructor(elements, completeAfterMs) {
+    this.completeAfterMs = completeAfterMs;
+    this.elementViewings = new Set();
+    elements.forEach((el) => {
+      this.elementViewings.add(new ElementViewing(el, completeAfterMs));
+    });
+    this.visible = new Set();
+    this.previouslyVisible = new Set();
+    this.handlers = [];
+    this.interval = undefined;
+    this.registerDomHandlers();
+  }
+
+  /** Register a new handler to be called when an element has been viewed.  */
+  addHandler(handler) {
+    this.handlers.push(handler);
+  }
+
+  windowHeight() { // eslint-disable-line class-methods-use-this
+    return window.innerHeight;
+  }
+
+  updateVisible() {
```

Comment:
Please add a comment explaining the purpose of this method as well as `triggerUpdates`.

Reply:
Also, the code in `updateVisible` and `triggerUpdates` is pretty hard to follow, and seems like it could be fragile. It's hard to tell from reading it how well it will account for edge cases like scrolling really fast past many small XBlocks while `completeAfterMs` is < 1000 ms.

May I suggest another approach? I got a bit carried away (sorry!) but this is still a sketch, so I'm not sure if it has the same functionality. I think something like this can avoid the need for tracking changes using `Set`s and the entire `triggerUpdates` method plus its interval toggle.

```javascript


class ElementViewing {
    constructor(el, completeAfterMs, callback) {
        ...
    }
    ...
    /** This element has become visible on screen (may be called even when already on screen though) */
    handleVisible() {
        if (!this.becameVisibleAt) {
            this.becameVisibleAt = Date.now();
            // We're now visible; after completeAfterMs, if the top and bottom have been
            // seen, this block will count as viewed.
            setTimeout(() => this.checkIfViewed(), this.completeAfterMs - this.seenForMs);
        }
    }
    handleNotVisible() {
        if (this.becameVisibleAt) {
            this.seenForMs = Date.now() - this.becameVisibleAt;
        }
        this.becameVisibleAt = null;
    }
    markTopSeen() {
        this.topSeen = true;
        checkIfViewed(); // If this element has been seen for enough time, but the bottom wasn't visible, it may now be viewed.
    }
    markBottomSeen() {
        this.bottomSeen = true;
        checkIfViewed();
    }
    checkIfViewed() {
        if (this.hasBeenViewed) {
            return;
        }
        const totalTimeSeen = this.seenForMs + this.becameVisibleAt ? (Date.now() - this.becameVisibleAt) : 0;
        if (this.topSeen && this.bottomSeen && totalTimeSeen >= completeAfterMs) {
            this.callback(this.el, {elementHasBeenViewed: true});
            this.hasBeenViewed = true;
        }
    }
}

class ViewedEventTracker {
    ...
    constructor(elements, completeAfterMs) {
        ...
        elements.forEach((el) => {
            this.elementViewings.add(new ElementViewing(el, completeAfterMs, (el, event) => this.callHandlers(el, event)));
        });
        ...
    }
    ...
    /** whenever the viewport changes, update each ElementViewing's visible status */
    updateVisible() {
        this.elementViewings.forEach((elv) => {
            if (elv.hasBeenViewed) {
                return;
            }
            const rect = elv.getBoundingRect();
            let onScreen = false;
            if (rect.top > 0 && rect.top < height) {
                elv.markTopSeen();
                onScreen = true;
            }
            if (rect.bottom > 0 && rect.bottom < height) {
                elv.markBottomSeen();
                onScreen = true;
            } else if (rect.top < 0 && rect.bottom > height) {
                onScreen = true;
            }
            if (onScreen) {
                elv.handleVisible();
            } else {
                elv.handleNotVisible();
            }
        }
    }
}
```

Reply:
Thanks.  These are helpful.  I've replied to all of them except the big refactor suggestion.  I'll tackle that in the morning.

Reply:
Great! Don't consider the refactor mandatory, but take a read and see if you consider it helpful or not. If not, I won't be bothered :)

Reply:
Overall, I think it looks clean.  One issue, though is that the element doesn't have access to the callbacks.  So I'll need to either pass it along (it can get added after blocks are already viewed), or have the blocks report to the ViewedTracker that they are done.

Reply:
Yup. In the example above, I showed the latter suggestion - passing a new `callHandlers` method as the callback to each `ElementViewing`.

So `ElementViewing` can always call its own `callback` which is a reference to `ViewedEventTracker.callHandlers()` which in turn calls whatever handlers have been registered into `ViewedEventTracker.handlers`.

But there's any number of ways to achieve that, and they're all more or less equivalent.